miden-crypto/src/encryption/xchacha/mod.rs

```diff
@@ -0,0 +1,419 @@
+    }
+
+    fn encrypt(
+        key: &Self::Key,
+        nonce: Self::Nonce,
+        plaintext: &[Felt],
+        associated_data: &[Felt],
+    ) -> Result<Vec<Felt>, IesError> {
+        let encrypted_data = key.encrypt_with_nonce(plaintext, associated_data, nonce)?;
+        Ok(encrypted_data.as_felts())
+    }
+
+    fn decrypt(
+        key: &Self::Key,
+        nonce: Self::Nonce,
+        ciphertext: &[Felt],
+        associated_data: &[Felt],
+    ) -> Result<Vec<Felt>, IesError> {
+        let encrypted_data = &EncryptedDataFelt {
+            associated_data: associated_data.to_vec(),
+            ciphertext: ciphertext.to_vec(),
+            nonce: bytes_to_felts(nonce.inner.as_slice()).try_into().unwrap(),
+        };
+        let decrypted_data = key.decrypt(encrypted_data)?;
+        Ok(decrypted_data)
+    }
+
+    fn encrypt_bytes(
+        key: &Self::Key,
+        nonce: Self::Nonce,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Vec<u8>, IesError> {
+        let encrypted_data = key.encrypt_bytes_with_nonce(plaintext, associated_data, nonce)?;
+        Ok(encrypted_data.to_bytes())
+    }
+
+    fn decrypt_bytes(
+        key: &Self::Key,
+        nonce: Self::Nonce,
+        ciphertext: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Vec<u8>, IesError> {
+        let encrypted_data = &EncryptedData {
+            associated_data: associated_data.to_vec(),
+            ciphertext: ciphertext.to_vec(),
+            nonce,
+        };
+        let decrypted_data = key.decrypt_bytes(encrypted_data)?;
+        Ok(decrypted_data)
+    }
+}
+
+// SERIALIZATION / DESERIALIZATION
+// ================================================================================================
+
+impl Serializable for SecretKey {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_bytes(&self.0);
+    }
+}
+
+impl Deserializable for SecretKey {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let inner: [u8; 32] = source.read_array()?;
+
+        Ok(SecretKey(inner))
+    }
+}
+
+impl Serializable for Nonce {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_bytes(self.inner.as_slice());
+    }
+}
+
+impl Deserializable for Nonce {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let inner: [u8; 32] = source.read_array()?;
+
+        Ok(Nonce {
+            inner: chacha20poly1305::XNonce::clone_from_slice(&inner),
+        })
+    }
+}
+
+impl Serializable for EncryptedData {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let associated_data_len = self.associated_data.len() as u32;
+        let ciphertext_len = self.ciphertext.len() as u64;
+
+        target.write_u32(associated_data_len);
+        target.write_bytes(&self.associated_data);
+        target.write_u64(ciphertext_len);
+        target.write_bytes(&self.ciphertext);
```

Comment:
I think we can use `write_usize()` instead of `write_u32()` and `write_u64()` here.